### ISSUE TITLE
关于VSD中KL散度计算的一点疑问

### ISSUE THREAD

**Issue by caixincx** [FutabaSakuraXD/Farewell-to-Mutual-Information-Variational-Distiilation-for-Cross-Modal-Person-Re-identification#5]
作者您好，您在VSD中计算了p(y|v)和p(y|z)的散度，计算的代码如下：
```
 vsd_loss = kl_div(input=self.softmax(i_observation[0].detach() / self.args.temperature),
                          target=self.softmax(i_representation[0] / self.args.temperature))
```
但是根据我的理解，p(y|v)和p(y|z)的KL散度应该是如下计算：
```
vsd_loss = kl_div(input=torch.nn.LogSoftmax(dim=1)(i_representation[0] / self.args.temperature),
                          target=self.softmax(i_observation[0].detach() / self.args.temperature))
```
请问为什么您在代码中是使用第一种实现方式？这两种实现方式是等价的吗？

**Comment by Qinying-Liu** [comment]
我也感觉很奇怪 第一种形式不仅形式上不对，而且感觉完全没有梯度

**Comment by caixincx** [comment]
> 我也感觉很奇怪 第一种形式不仅形式上不对，而且感觉完全没有梯度
是的，我做实验发现这一项的loss很小，几乎没啥影响

**Comment by FutabaSakuraXD** [comment]
Check the following contents before using the KL divergence in pytorch:


https://zhuanlan.zhihu.com/p/575809052?utm_id=0



https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss


Or try with ur own formats if you want
: )

--------------原始邮件--------------
发件人："caixincx ***@***.***&gt;;
发送时间：2022年11月23日(星期三) 中午11:57
收件人："FutabaSakuraXD/Farewell-to-Mutual-Information-Variational-Distiilation-for-Cross-Modal-Person-Re-identification" ***@***.***&gt;;
抄送："Subscribed ***@***.***&gt;;
主题：Re: [FutabaSakuraXD/Farewell-to-Mutual-Information-Variational-Distiilation-for-Cross-Modal-Person-Re-identification] 关于VSD中KL散度计算的一点疑问 (Issue #5)
-----------------------------------

 

  
我也感觉很奇怪 第一种形式不仅形式上不对，而且感觉完全没有梯度
 是的，我做实验发现这一项的loss很小，几乎没啥影响

**Comment by Qinying-Liu** [comment]
Hi FutabaSakuraXD. Thank you for your reply. For the standard form of _torch.nn.KLDivLoss_ or _F.kl_div_, I think that we have to first perform a _log_ operation on the inputs.   
![image](https://user-images.githubusercontent.com/33945655/203471666-b7c0b2c3-5c8b-4687-94d1-c6de9b8c906d.png)
Moreover, the grad of the inputs should not be detached.

**Comment by FutabaSakuraXD** [comment]
Gradients of the observation are detached to prevent degeneration.

**Comment by canbaoburen** [comment]
Then there will be not any gradient as the torch.nn.KLDivLoss or F.kl_div will automatically detach the gradients of the target while your input has already been detached as far as I know.
